### PR TITLE
onDidExecuteCommand API

### PR DESCRIPTION
**Pull request by hedgerh**
Implemented based on conversations in https://github.com/Microsoft/vscode/issues/1431

This PR adds a new `onDidExecuteCommand` event to the `commands` namespace of the VSCode API. 
 This allows extensions to be able to subscribe and react to both extension and UI side events with a callback function.

One possible use case is a live macro recording feature.

Example usage:
```ts
vscode.commands.onDidExecuteCommand(({ commandId, args }: { commandId: string, args: any[] }) => {
  if (commandId === 'type') {
    vscode.window.showInformationMessage(`Key pressed: ${args[0]}`)
  }
})
```

### CODE REVIEW

File: src/vs/vscode.proposed.d.ts [hunk 1]
```diff
@@ -1,3 +1,5 @@
+import { ICommandEvent } from 'vs/platform/commands/common/commands'
```
Comment (jrieken) [inline src/vs/vscode.proposed.d.ts]: No dependencies allowed here, the Event type must be defined again, e.g. `interface ExecuteCommandEvent { command:string, args?:any[] }`

File: src/vs/workbench/api/common/extHostCommands.ts [hunk 2]
```diff
@@ -30,6 +32,8 @@ export interface ArgumentProcessor {
 }
 
 export class ExtHostCommands implements ExtHostCommandsShape {
+	private readonly _onDidExecuteCommand: Emitter<ICommandEvent> = new Emitter<ICommandEvent>();
+	public readonly onDidExecuteCommandEmitter: Event<ICommandEvent> = this._onDidExecuteCommand.event;
```
Comment (jrieken) [inline src/vs/workbench/api/common/extHostCommands.ts]: No `public` needed

File: src/vs/workbench/api/common/extHostCommands.ts [hunk 3]
```diff
@@ -133,7 +146,9 @@ export class ExtHostCommands implements ExtHostCommandsShape {
 				}
 			});
 
-			return this._proxy.$executeCommand<T>(id, args).then(result => revive(result, 0));
+			return this._proxy.$executeCommand<T>(id, args).then(result => {
+				return revive(result, 0);
+			});
```
Comment (jrieken) [inline src/vs/workbench/api/common/extHostCommands.ts]: Unnecessary change?

File: src/vs/workbench/api/node/extHost.api.impl.ts [hunk 4]
```diff
@@ -230,7 +230,10 @@ export function createApiFactory(
 			},
 			getCommands(filterInternal: boolean = false): Thenable<string[]> {
 				return extHostCommands.getCommands(filterInternal);
-			}
+			},
+			onDidExecuteCommand: (listener, thisArgs?, disposables?) => {
+				return extHostCommands.onDidExecuteCommand(listener, thisArgs, disposables);
```
Comment (jrieken) [inline src/vs/workbench/api/node/extHost.api.impl.ts]: Check if proposed API is allows for this extensions, see: https://github.com/Microsoft/vscode/blob/84b1474d097789b7915445688b2201f9d897d0cf/src/vs/workbench/api/node/extHost.api.impl.ts#L501

File: src/vs/workbench/services/commands/common/commandService.ts [hunk 5]
```diff
@@ -77,8 +80,12 @@ export class CommandService extends Disposable implements ICommandService {
 			return Promise.reject(new Error(`command '${id}' not found`));
 		}
 		try {
-			this._onWillExecuteCommand.fire({ commandId: id });
+			this._onWillExecuteCommand.fire({ commandId: id, args });
 			const result = this._instantiationService.invokeFunction.apply(this._instantiationService, [command.handler, ...args]);
+			this._onDidExecuteCommand.fire({
+				commandId: id,
+				args,
+			});
```
Comment (jrieken) [inline src/vs/workbench/services/commands/common/commandService.ts]: inconsistent formatting, see `_onWill`...

File: src/vs/workbench/api/common/extHostCommands.ts [hunk 6]
```diff
@@ -107,6 +111,15 @@ export class ExtHostCommands implements ExtHostCommandsShape {
 		});
 	}
 
+	onDidExecuteCommand(listener: (command: ICommandEvent) => any, thisArgs?: any, disposables?: IDisposable[]) {
+		this._proxy.$onDidExecuteCommand();
+		return this.onDidExecuteCommandEmitter(listener, thisArgs, disposables);
```
Comment (jrieken) [inline src/vs/workbench/api/common/extHostCommands.ts]: Not sure what this is about? Isn't this just re-exposing the event as function? Above exists `onDidExecuteCommandEmitter` (which isn't an emitter but an event btw) which should be enough. Also, I don't understand why the `$onDidExecuteCommand` call is needed? The main side only needs one listener for main-side events which it then forwards to the ext host side.

File: src/vs/workbench/api/common/extHostCommands.ts (hunk 3 above)
Comment (jrieken) [reply src/vs/workbench/api/common/extHostCommands.ts]: The `executeCommand` function checks if that command itself is contributed by an extension. In that case the execution stays within this process and you need to manually create/fire an event

**Review by jrieken — CHANGES_REQUESTED** [review]

File: src/vs/workbench/api/common/extHostCommands.ts (hunk 6 above)
Comment (hedgerh) [reply src/vs/workbench/api/common/extHostCommands.ts]: are you referring to the return statement?  I got a little confused about what I should be returning from `onDidExecuteCommand`.  Most of the events seem to return a `Disposable`, so I was just trying to copy patterns.  Probably way off there.

With the `$onDidExecuteCommand` call, my thinking was that I wouldn't register a subscriber to listen to events on the main-side until an extension actually asked for it by calling `onDidExecuteCommand`.  Does it make more sense to just have one listener always on the main side that forwards events over, even if no one is listening on the extension side?

Comment (jrieken) [reply src/vs/workbench/api/common/extHostCommands.ts]: > With the $onDidExecuteCommand call, my thinking was that I wouldn't register a subscriber to listen to events on the main-side until an extension actually

That makes sense but now you add a listener on the main side for each listener on the extension side. When instantiating an `Emitter` you can provide options with callbacks that are invoked for the first and last listener. That's what you wanna use.

File: src/vs/workbench/api/common/extHostCommands.ts (hunk 3 above)
Comment (hedgerh) [reply src/vs/workbench/api/common/extHostCommands.ts]: Fixing formatting.

The contributed commands are actually passing through the `commandService`, so events are firing for them.

Comment (jrieken) [reply src/vs/workbench/api/common/extHostCommands.ts]: Not those: https://github.com/Microsoft/vscode/blob/23fd99b493979a5b3778dc239d7eea46bf249f16/src/vs/workbench/api/common/extHostCommands.ts#L116

Try with `vscode.commands.executeCommand` and call a command your or another extension contributes.

Comment (hedgerh) [reply src/vs/workbench/api/common/extHostCommands.ts]: Right right, those are the ones I was referring to.  `extension.helloWorld` seems to already get captured with no changes to the code.  After adding a manual fire after the contributed command is executed in `_executeContributedCommand`, it now fires off twice.  Not 100% sure how, to be honest.

Comment (jrieken) [reply src/vs/workbench/api/common/extHostCommands.ts]: It matters how you invoke the command. User gestures like the command-palette or keybindings are happening via the main side should be covered but just calling `executeCommand` (e.g from a timer) is different

File: src/vs/workbench/api/common/extHostCommands.ts (hunk 6 above)
Comment (hedgerh) [reply src/vs/workbench/api/common/extHostCommands.ts]: Sorry, can you walk me through what you're thinking?  

Should the main-side listener be added only the first time an extension subscribes to the event, instead of having a one-to-one ratio with the extension side?  How would the first/last callbacks play into it?

Comment (jrieken) [reply src/vs/workbench/api/common/extHostCommands.ts]: Yes, there is a single listener on the main side that forwards events to extension side. The extension side will have its own emitter/event to which extensions subscribe. That makes sure only one IPC message is send per event, not one per listener, and only if at least one listener exsits

File: src/vs/workbench/api/node/extHost.api.impl.ts (hunk 4 above)
Comment (hedgerh) [reply src/vs/workbench/api/node/extHost.api.impl.ts]: Added.

File: src/vs/workbench/api/browser/mainThreadCommands.ts [hunk 7]
```diff
@@ -78,9 +78,21 @@ export class MainThreadCommands implements MainThreadCommandsShape {
 		return this._commandService.executeCommand<T>(id, ...args);
 	}
 
+	$onDidExecuteCommand() {
+		return this._commandService.onDidExecuteCommand((command) => this.handleExecuteCommand(command));
+	}
+
+	$disposeListeners() {
+		return this._commandService.disposeListeners();
```
Comment (hedgerh) [inline src/vs/workbench/api/browser/mainThreadCommands.ts]: Added `$disposeListeners` to call the `dispose` on the main thread event emitter when there are no more listeners subscribed on the extension side.  Seem legit?

File: src/vs/workbench/api/common/extHostCommands.ts (hunk 6 above)
Comment (hedgerh) [reply src/vs/workbench/api/common/extHostCommands.ts]: Moved the proxy calls into the `onFirstListenerDidAdd` and `onLastListenerRemove` hooks on the ext Emitter.

One thing:  I need to read up on Disposables a bit more.  I *think* I'm supposed to return `this.onDidExecuteCommand` here, so extensions get a `disposable` that they can push to `context.subscriptions`?  I'm not 100% here.  hopefully it's right.

File: src/vs/workbench/api/common/extHostCommands.ts (hunk 3 above)
Comment (hedgerh) [reply src/vs/workbench/api/common/extHostCommands.ts]: Added.

**Comment by hedgerh** [comment]
This is ready for review when you have the chance.  I left some comments on a few discussion points that I wanted to get your sign off on, but they should hopefully be resolvable.

File: src/vs/workbench/services/commands/common/commandService.ts [hunk 8]
```diff
@@ -22,6 +22,10 @@ export class CommandService extends Disposable implements ICommandService {
 	private readonly _onWillExecuteCommand: Emitter<ICommandEvent> = this._register(new Emitter<ICommandEvent>());
 	public readonly onWillExecuteCommand: Event<ICommandEvent> = this._onWillExecuteCommand.event;
 
+	private readonly _onDidExecuteCommand: Emitter<ICommandEvent> = new Emitter<ICommandEvent>();
+	public readonly onDidExecuteCommand: Event<ICommandEvent> = this._onDidExecuteCommand.event;
+	public readonly disposeListeners: () => void = () => this._onDidExecuteCommand.dispose();
```
Comment (jrieken) [inline src/vs/workbench/services/commands/common/commandService.ts]: Remove `disposeListeners`

File: src/vs/editor/test/browser/editorTestServices.ts [hunk 9]
```diff
@@ -43,7 +48,7 @@ export class TestCommandService implements ICommandService {
 		}
 
 		try {
-			this._onWillExecuteCommand.fire({ commandId: id });
+			this._onWillExecuteCommand.fire({ commandId: id, args });
```
Comment (jrieken) [inline src/vs/editor/test/browser/editorTestServices.ts]: `_onDidExec` not being fired...

File: src/vs/workbench/api/browser/mainThreadCommands.ts (hunk 7 above)
Comment (jrieken) [reply src/vs/workbench/api/browser/mainThreadCommands.ts]: No, this is not how it works. Your implementation of `CommandService#disposeListeners` removes all listeners and kills the `CommandService#onDidExecuteCommand` event. 

* `ICommandService#onDidExecuteCommand` is an even with potentially *many* listeners.
* `MainThreadCommands` can be a (1**!**) listener for such event. Its implementation forwards the event data to the extension host process
* The extension host process (`extHostCommands`) receives the event data and all it does is firing another event. That event (`Emitter`) is exposed as API and manages its own set of listeners. That means when an "API-listener" stops listening nothing needs to change in the `ICommandService`

**Review by jrieken — CHANGES_REQUESTED** [review]

File: src/vs/workbench/api/browser/mainThreadCommands.ts (hunk 7 above)
Comment (hedgerh) [reply src/vs/workbench/api/browser/mainThreadCommands.ts]: Got it.  To be clear, should I still add/kill the listener on first add/last remove on the `Emitter` in `extHostCommands`, but only get rid of the listener in `MainThreadCommands`?  

```
this._onDidExecuteCommand = new Emitter<ICommandEvent>({
	onFirstListenerDidAdd: this._proxy.$onDidExecuteCommand,
	onLastListenerRemove: this._proxy.$disposeListener,
});
```

I assume that's preferable to having the listener in `MainThreadCommands` always forward events despite no listeners in the extension side `Emitter`, but just want to run it by you first.  Thanks for the assistance, btw.

Comment (hedgerh) [reply src/vs/workbench/api/browser/mainThreadCommands.ts]: Just updated the code to add/dispose of the single listener

Comment (jrieken) [reply src/vs/workbench/api/browser/mainThreadCommands.ts]: Yeah, think of it (and tackle it) as two problems. 

1) to implement the `onDidExecuteCommand` on the extension host side you need the forwarding so that you can drive an `Emitter` there
2) now when no-one listens on the extension host side you still forward all those events for nothing, wasting IPC messages/resource. that's where the added first, removed last listener comes into play.

Comment (hedgerh) [reply src/vs/workbench/api/browser/mainThreadCommands.ts]: Got it.  Just wanted to confirm.  I pushed up changes to add/remove only the single listener in `MainThreadCommands`.  Should be good to go.

File: src/vs/workbench/api/common/extHost.protocol.ts [hunk 10]
```diff
@@ -113,6 +113,8 @@ export interface MainThreadClipboardShape extends IDisposable {
 
 export interface MainThreadCommandsShape extends IDisposable {
 	$registerCommand(id: string): void;
+	$onDidExecuteCommand(): void;
+	$disposeListener(): void;
```
Comment (jrieken) [inline src/vs/workbench/api/common/extHost.protocol.ts]: For clarity I would call these two `registerCommandListener` and `unregisterCommandListener`

File: src/vs/workbench/api/common/extHostCommands.ts [hunk 11]
```diff
@@ -43,6 +47,11 @@ export class ExtHostCommands implements ExtHostCommandsShape {
 		logService: ILogService
 	) {
 		this._proxy = mainContext.getProxy(MainContext.MainThreadCommands);
+		this._onDidExecuteCommand = new Emitter<ICommandEvent>({
+			onFirstListenerDidAdd: this._proxy.$onDidExecuteCommand,
+			onLastListenerRemove: this._proxy.$disposeListener,
```
Comment (jrieken) [inline src/vs/workbench/api/common/extHostCommands.ts]: Not sure but those two might miss a bind-call or change to `onFirstListenerDidAdd: () => this._proxy.$onDidExecuteCommand()` (and corresponding)

**Review by jrieken — CHANGES_REQUESTED** [review]
:+1: looks much better now, left a few nits but then it'll be good for merging

**Review by jrieken — APPROVED** [review]

**Comment by jrieken** [comment]
Thanks! This now looks good to me. We are currently in endgame-mode and I have to see if I can still merge this for the April release, otherwise this goes in for May

**Comment by hedgerh** [comment]
Hey hey @jrieken, just checking in on this PR.  Is there anything I should be doing at this point?  I read the docs on proposing Extension API changes, but it sounds like it's geared towards actual MS team members.  Do you all pick carry the torch from here?  Cheers.

**Comment by jrieken** [comment]
Yeah, sorry. I keep moving this as I haven't done proper testing about the affects of sending arguments over yet. Internally we often send arguments that cannot be serialised, e.g. we will fail to send them to the extension host side. While that's fine we don't know of the (performance) implications yet.
